### PR TITLE
Add reconciliation loop for job resource

### PR DESCRIPTION
**Pull request by hlts2**
### Description:

we (@kevindiu @hlts2 ) added reconciliation loop logic for job resource.

### Environment:

- Go Version: 1.15.2
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.1

### Types of changes:

- [ ] Bug fix [type/bug]
- [x] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by kevindiu — APPROVED** [review]
LGTM

**Review by kevindiu — APPROVED** [review]

**Review by kevindiu — APPROVED** [review]

File: internal/k8s/metrics/node/node.go [hunk 1]
```diff
@@ -61,7 +61,7 @@ func New(opts ...Option) NodeWatcher {
 }
 
 func (r *reconciler) Reconcile(req reconcile.Request) (res reconcile.Result, err error) {
-	m := &metrics.NodeMetricsList{}
+	m := new(metrics.NodeMetricsList)
```
Comment (kmrmt) [inline internal/k8s/metrics/node/node.go:64]: It is not essential, but I think this fix and the similar ones should be in the other PR.
How about it?

File: internal/k8s/job/job.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	mgr         manager.Manager
+	name        string
+	namespance  string
+	onError     func(err error)
+	onReconcile func(jobList map[string][]Job)
+}
+
+// Job represents k8s job imformation
+// https://godoc.org/k8s.io/api/batch/v1#Job
+type Job struct {
```
Comment (rinx) [inline internal/k8s/job/job.go:36]: Why it is not a type alias for `k8s.io/api/batch/v1.Job`? it is better to use that one if this package is an internal package that is used generally in various applications.

Comment (kevindiu) [reply internal/k8s/job/job.go:36]: @kpango what do you think about this?

Comment (kpango) [reply internal/k8s/job/job.go:36]: In the Reconcile loop of the pod package, I had to calculate the CPU/memory requirements and limits of the pod, so I defined a specific type (not an alias for the type), but in the Job package I don't think we need to calculate it, so we can remove the Job struct and  add the type alias or use the directory reference of `k8s.io/api/batch/v1.Job`.

Comment (kevindiu) [reply internal/k8s/job/job.go:36]: FYI all of the current implementation of reconciler has the same problem/question 
e.g.
https://github.com/vdaas/vald/blob/master/internal/k8s/node/node.go#L91

Comment (kpango) [reply internal/k8s/job/job.go:36]: In node package and pod package we need to define specific structure type not type alias.

File: internal/k8s/metrics/node/node.go (hunk 1 above)
Comment (hlts2) [reply internal/k8s/metrics/node/node.go:64]: I think so, too.:bow:
But it's not a big change, so I don't think this change is a problem.
And kpango asked us to include this change in the same PR.

File: internal/k8s/job/job.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/vdaas/vald/internal/errors"
+	"github.com/vdaas/vald/internal/k8s"
+	"github.com/vdaas/vald/internal/log"
+)
+
+// JobWatcher is is a type alias for k8s resource controller.
```
Comment (kevindiu) [inline internal/k8s/job/job.go:22]:
```suggestion
// JobWatcher is a type alias for k8s resource controller.
```

File: internal/k8s/job/job.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+				Requeue:      true,
+				RequeueAfter: time.Second,
+			}, nil
+		}
+		return
+	}
+
+	var (
+		jobs = make(map[string][]Job, len(js.Items))
+	)
```
Comment (kevindiu) [inline internal/k8s/job/job.go:75]:
```suggestion
	jobs := make(map[string][]Job, len(js.Items))
```

**Review by kevindiu — APPROVED** [review]
LGTM

**Review by vankichi — APPROVED** [review]

**Pull request merged**